Permutive RTD module: support IAB Audience taxonomy

desbo opened the pull request:
## Type of change
- [x] Feature

## Description of change
Updates the Permutive RTD module to facilitate for segmentation by
the new IAB Audience taxonomy.

To achieve this, this change introduces the concept of "transformations"
on the ORT2B `user.data` object. There are two components to these
transformations: a new `transformations` property on the Prebid config,
to be set by the publisher, and logic in the module for the actual
behaviour of the transformation.

We plan to use the transformation logic in this PR, combined with
configuration we'll share with publishers, to send IAB Audience
taxonomy cohort IDs to bidders.

patmmccann commented:
This feels like a super clunky experience for publishers. Why doesn't your endpoint return the transformation mappings instead of asking publishers to define potentially enormous config?

patmmccann commented on integrationExamples/gpt/permutiveRtdProvider_example.html:
```diff
@@ -160,7 +178,7 @@
               });
               pbjs.setBidderConfig({
-                bidders: ['appnexus', 'rubicon'],
+                bidders: ['appnexus', 'rubicon', 'ix'],
```
Please fix line 202 below, this syntax is out of date

patmmccann commented on modules/permutiveRtdProvider.js:
```diff
@@ -260,6 +272,34 @@ function readSegments (key) {
+const ortb2UserDataTransformations = {
+  iabAudienceTaxonomy11: (userData, config) => ({
+    name: userData.name,
+    ext: { segtax: '4' },
```
If you're making pubs define all the transformations, no reason you can't let this be a param...

patmmccann commented on test/spec/modules/permutiveRtdProvider_spec.js:
```diff
@@ -51,6 +51,45 @@ describe('permutiveRtdProvider', function () {
+          },
+          {
+            name: 'permutive.com',
+            ext: { segtax: '4' },
```
It's an integer

patmmccann commented on test/spec/modules/permutiveRtdProvider_spec.js:
```diff
@@ -51,6 +51,45 @@ describe('permutiveRtdProvider', function () {
+            segment: expectedTargetingData
+          },
+          {
+            name: 'permutive.com',
```
Is it really permutive that decided the audience member belongs in that iab segment if permutive fails to provide the transformations? The publisher could easily transform any permutive segment to any iab segment; so I'm not sure this remains correct.

patmmccann commented on modules/permutiveRtdProvider.js:
```diff
@@ -84,23 +85,34 @@ export function setBidderRtb (auctionDetails, customModuleConfig) {
-  const segment = segmentData.ac.map(seg => {
-    return { id: seg }
-  })
+function mergeOrtbConfig (currConfig, segmentIDs, transformationConfigs) {
```
There's a utils finding for this now, i think it's called mergeDeep

desbo replied:
Thanks for the suggestion – after looking at this function I realised it's not really doing a simple merge of two objects but actually updating the `user.data` array with some new entries. So I don't think `mergeDeep` is suitable here, but I've renamed the function to clarify what it does.

desbo commented:
@patmmccann thanks for the review so far. In terms of the overhead for publishers: I should have mentioned that this change is to support a test with a single publisher that will only involve a small selection of cohorts. For this test, Permutive will provide the publisher with the segment mapping for use in their config, and we already have agreement from the publisher on this way of working for the test.

Our plan was to go with this approach for the test and then revisit our implementation so that the segment mappings come from our SDK rather than the Prebid module in the long term. Ultimately I'd expect that publishers would only need to include a simple flag in their config. Would you accept these changes in light of this extra context?

patmmccann commented:
If you make segtax default to 4 but be overridable in config; i'd be happy to merge. Publishers want code that is "future-proof"; they don't want to have to suddenly upgrade prebid to support segtax i, j or k in their permutive adapter. We much prefer when adapters keep this in mind.

desbo commented:
@patmmccann thanks! I've addressed your comments now, most importantly to include `segtax` as a config parameter. 
Eventually we'll be back with another PR that removes the need for publishers to include the segment ID mappings in their config too, but hopefully this is acceptable in its current form for the test I mentioned.

patmmccann commented on integrationExamples/gpt/permutiveRtdProvider_example.html:
```diff
@@ -180,11 +199,6 @@
-                          name: 'www.dataprovider1.com',
-                          ext: { taxonomyname: 'iab_audience_taxonomy' },
-                          segment: [{ id: '687' }, { id: '123' }]
-                        },
```
two things left: circleci needs to run; add a segtax here. Can be the permutive contextual segtax or the iab contextual (don't use the audience bc that is user.data)

patmmccann commented:
> @patmmccann thanks! I've addressed your comments now, most importantly to include `segtax` as a config parameter. Eventually we'll be back with another PR that removes the need for publishers to include the segment ID mappings in their config too, but hopefully this is acceptable in its current form for the test I mentioned.
> 
> edit to mention that I'll be away for the next couple of weeks and my colleagues @bartholomews and @Mullefa will commit to this branch if there are any more changes to make.

Thanks so much! Did you notice circleci didnt run? Usually helps to unfollow your fork in circleci before you make the change. Just one tiny change in the example file requested.

desbo commented:
@patmmccann thanks again. I've added `segtax` to the example, hopefully in the right place but if not, please let me know and one of my teammates will fix. CircleCI is now running after updating this branch with master, although a lot of the tests are timing out – think these are unrelated to my changes?

bartholomews commented:
Hi @patmmccann circleci is now green, please let me know if there's anything else left to address! Thanks
